### ISSUE TITLE
flaky test: TestFileCache_{compiler, wazevo}

### ISSUE THREAD

**Issue by mathetake** [tetratelabs/wazero#2039]
We recently are having the flaky seg fault as in 

* https://github.com/tetratelabs/wazero/actions/runs/7853356149/job/21432882550?pr=2036
* https://github.com/tetratelabs/wazero/actions/runs/7856559936/job/21439463102

since the failure is happening for both wazevo and old compiler, I am pretty confident that 
this is something to do with life cycle issue around mmap/finalizer in file caches.

**Comment by mathetake** [comment]
relevant: https://github.com/tetratelabs/wazero/pull/1815

**Comment by evacchi** [comment]
Unrelated, but I am noticing that if the function returns early (line 36) then the finalizer is not set.

https://github.com/tetratelabs/wazero/blob/99c057bcb5f3f4311815c884ba7113c8d52b3316/internal/engine/compiler/engine_cache.go#L35-L54

As for the actual issue. I still don't know, but one thing that I notice is that before this commit the original code was a little more defensive see [here](https://github.com/tetratelabs/wazero/commit/9780f0f4a019ebf8c657fd74f65ce25e2046e507#diff-45ddcb29b654e977deee8ee51939c95ca6320460c241f4ace76116c316f43df6R544-R547) and [especially here](https://github.com/tetratelabs/wazero/commit/9780f0f4a019ebf8c657fd74f65ce25e2046e507#diff-45ddcb29b654e977deee8ee51939c95ca6320460c241f4ace76116c316f43df6L476-L485)

e.g. originally:

```
	e := cm.executable
	if e == nil {
		return // already released
	}

	// Setting this to nil allows tests to know the correct finalizer function was called.
	cm.executable = nil
	if err := platform.MunmapCodeSegment(e); err != nil {
		// munmap failure cannot recover, and happen asynchronously on the finalizer thread. While finalizer
		// functions can return errors, they are ignored.
		panic(fmt.Errorf("compiler: failed to munmap code segment: %w", err))
	}
```

and I wonder if the defensive check against nil was actually needed. WDYT @achille-roussel could this be related at all?

**Comment by achille-roussel** [comment]
My understanding of this code is that we don't need to set the finalizer on values we got from the in-memory table of compiled modules because they have already been attached a finalizer if needed (e.g., if they were obtained from the cache).

I think the current code we have still has the same validation, it's just been moved into the `asm.(*CodeSegment).Unmap` function https://github.com/tetratelabs/wazero/commit/9780f0f4a019ebf8c657fd74f65ce25e2046e507#diff-7257d290f32f0a9c96c45985d8131d1a37533bfb68bb6cff425cf11d15f25757R66-R75

```go
func (seg *CodeSegment) Unmap() error {
	if seg.code != nil {
		if err := platform.MunmapCodeSegment(seg.code[:cap(seg.code)]); err != nil {
			return err
		}
		seg.code = nil
		seg.size = 0
	}
	return nil
}
```

**Comment by evacchi** [comment]
I can reproduce the same error (although I don't know if am reproducing the exact same conditions) if I try to `releaseCompiledModule(cm)` while that module is being executed. Essentially:

```
		go modInst.ExportedFunction("1").Call(testCtx)
		releaseCompiledModule(cm)
```

So my guess is the finalizer might have to be guarded to check whether we are unmapping a region that is still in use because of some race condition:

1. we scheduled a region to be released
2. the module is no longer being executed/deleted, so eventually the GC will want to collect
3. the module is retrieved again from the file cache because we reloaded it concurrently
4. the finalizer is still scheduled to unmap, so it kills the mmap'd region while we are inside it

**Comment by evacchi** [comment]
more observations that might be useful. 

- The first failure is on [macOS amd64](https://github.com/tetratelabs/wazero/actions/runs/7853356149/job/21432882550?pr=2036) during a normal run of the test suite.
- The second failure is on [linux amd64](https://github.com/tetratelabs/wazero/actions/runs/7856559936/job/21439463102) but it is a **scratch container**. So the test is first **compiled** and then the container is created.

Now, because tests are compiled by package:

    # use CGO_ENABLED=0 to ensure the build does not link against glibc
    CGO_ENABLED=0 go test ./internal/integration_test/filecache/... -c -o filecache.test

it means that the executable has run for a relatively short time, thus it can't have accumulated "a lot" of garbage, and that garbage is only related to the tests in that sub-package. So this rules out any interference from "other" tests.

This is how I am running `filecache.test` similarly to how it runs on CI (case 2)

```
#Dockerfile
FROM scratch
CMD ["/test"]
```

For completeness, I am running my tests on Fedora using buildah/podman:

    buildah build -t wazero:test  .
    podman run --cpus=0.5 --cpuset-cpus=0,1 --memory=300m  -v $PWD/filecache.test:/test --tmpfs /tmp -it localhost/wazero:test /test
    
(--cpuset-cpus= [might need to be configured](https://docs.podman.io/en/latest/markdown/podman-run.1.html#cpuset-cpus-number))
    
and then I am creating a stresstest on the same cpus 

    taskset -c 0,1 stress --cpu 4 --vm 4 --vm-hang 1

**Comment by evacchi** [comment]
https://github.com/tetratelabs/wazero/pull/2091 follows up to #2088 and hardens the cache

**Comment by mathetake** [comment]
https://github.com/tetratelabs/wazero/actions/runs/8077516060/job/22067963967?pr=2099 this hasn't been resolved yet I think

**Comment by evacchi** [comment]
it's interesting that in both cases when the failure was in the old compiler it was amd64+macos-12 🧐

**Comment by evacchi** [comment]
I think I got it in https://github.com/tetratelabs/wazero/pull/2102 

tl;dr: there's a call_indirect in a spec test that points to a function that does not exist because its module has been collected earlier
